services/w3c/w3c-validation.service.js

```diff
@@ -0,0 +1,269 @@
+      namedLogo: 'w3c',
+    }
+  }
+
+  static render({ message, color }) {
+    return {
+      label: 'w3c',
+      message,
+      color,
+    }
+  }
+
+  getSchema(preset) {
+    const schema = []
+    const decodedPreset = decodeURI(preset)
+    if (decodedPreset.match(new RegExp(html4Expression, 'i'))) {
+      if (decodedPreset.match(/Strict/i)) {
+        schema.push('http://s.validator.nu/xhtml10/xhtml-strict.rnc')
+      } else if (decodedPreset.match(/Transitional/i)) {
+        schema.push('http://s.validator.nu/xhtml10/xhtml-transitional.rnc')
+      } else {
+        schema.push('http://s.validator.nu/xhtml10/xhtml-frameset.rnc')
+      }
+      schema.push('http://s.validator.nu/html4/assertions.sch')
+      schema.push('http://c.validator.nu/all-html4/')
+    } else {
+      if (decodedPreset.match(new RegExp(html5Expression, 'i'))) {
+        if (decodedPreset.match(/ITS 2\.0/i)) {
+          schema.push('http://s.validator.nu/html5-its.rnc')
+        } else if (decodedPreset.match(/RDFa Lite 1\.1/i)) {
+          schema.push('http://s.validator.nu/html5-rdfalite.rnc')
+        } else {
+          schema.push('http://s.validator.nu/html5.rnc')
+        }
+      } else if (decodedPreset.match(new RegExp(xhtmlExpression, 'i'))) {
+        if (decodedPreset.match(/RDFa Lite 1\.1/i)) {
+          schema.push('http://s.validator.nu/xhtml1-ruby-rdf-svg-mathml.rnc')
+        } else if (decodedPreset.match(/1\.0 Strict, URL, Ruby, SVG 1\.1/i)) {
+          schema.push('http://s.validator.nu/xhtml1-ruby-rdf-svg-mathml.rnc')
+        } else {
+          schema.push('http://s.validator.nu/xhtml5.rnc')
+        }
+      } else if (decodedPreset.match(new RegExp(svgExpression, 'i'))) {
+        schema.push('http://s.validator.nu/svg-xhtml5-rdf-mathml.rnc')
+      }
+      schema.push('http://s.validator.nu/html5/assertions.sch')
+      schema.push('http://c.validator.nu/all/')
+    }
+    return schema.join(' ')
+  }
+
+  getMessage(messageTypes) {
+    const messageTypeKeys = Object.keys(messageTypes)
+    messageTypeKeys.sort() // Sort to make the order error, warning for display
+
+    if (Object.keys(messageTypes).length === 0) {
+      return 'validated'
+    }
+
+    const messages = messageTypeKeys.map(
+      key => `${messageTypes[key]} ${key}${messageTypes[key] > 1 ? 's' : ''}`
+    )
+    return messages.join(', ')
+  }
+
+  getColor(messageTypes) {
+    if ('error' in messageTypes) {
+      return 'red'
+    }
+
+    if ('warning' in messageTypes) {
+      return 'yellow'
+    }
+
+    return 'brightgreen'
+  }
+
+  getUrl(doc, preset, parser) {
+    const qs = { doc, out: 'json' }
+    if (preset) qs['schema'] = this.getSchema(preset)
+    if (parser) qs['parser'] = parser
+    return {
+      url: 'http://validator.nu',
+      schema,
+      options: {
+        qs,
+      },
+    }
+  }
+
+  async fetch(jsonUrl) {
+    return this._requestJson(jsonUrl)
+  }
+
+  async handle(_routeParams, { doc, preset, parser }) {
```

Comment:
```suggestion
  async handle({ preset }, { doc, parser }) {
```

Let's also refactor a lot of the content in this function out based on responsibilities. For example, a really common pattern we use is to have a `transform` function that takes the raw API response (which would be `data` here) and inspecting/analyzing that response to return the results.

There's also no need to have `getUrl` as a separate function that's called from `handle`, let's move all of that into the `fetch` function. 

That would then result in `handle` looking something like this: 
```suggestion
  async handle({ preset }, { url, parser }) {
    const data = await this.fetch({ url, preset, parser })
    const messages = this.transform(data)
    ...
```

Reply:
New signature `async handle({ parser }, { targetUrl, preset }) {`